### PR TITLE
Python: Yield FunctionResultContent in streaming chat completion path. Update tests.

### PR DESCRIPTION
**Pull request by moonbox3**
### Motivation and Context

Currently, if using SK's Python streaming chat completion path, we only yield the following content types: `StreamingChatMessageContent` and `FunctionCallContent`. We are not yielding `FunctionResultContent` which is valuable for some use cases.

### Description

This PR updates the code to yield `FunctionResultContent` if it exists in the streaming chat completion path. When we merge the function call result content together into a `StreamingChatMessageContent` type, we check if that message has items (which are of type `FunctionResultContent`) and if so, we yield them. The filter path still works because once they're yielded, we break out of the function calling loop.

We need to include the `ai_model_id` if it exists for the current PromptExecutionSettings because if performing a `reduce` operation to add two streaming chat message chunks together, the `StreamingChatMessageContent` that has the function results will break if the `ai_model_id` is not set (the error is thrown in the `__add__` override for the `StreamingChatMessageContent`.

Some unit tests that cover function calling were also updated -- during the test, the test JSON function args were breaking in the `json.loads` call because they contained single quotes and not double. We're now sanitizing the args, just in case, so we don't break there. 

This PR fixes:
- #9408
- #9968 

### Contribution Checklist

- [X] The code builds clean without any errors or warnings
- [X] The PR follows the [SK Contribution Guidelines](https://github.com/microsoft/semantic-kernel/blob/main/CONTRIBUTING.md) and the [pre-submission formatting script](https://github.com/microsoft/semantic-kernel/blob/main/CONTRIBUTING.md#development-scripts) raises no violations
- [X] All unit tests pass, and I have added new tests where possible
- [X] I didn't break anyone :smile:

### CODE REVIEW

File: python/semantic_kernel/connectors/ai/chat_completion_client_base.py [hunk 1]
```diff
@@ -303,8 +306,18 @@ async def get_streaming_chat_message_contents(
...
+                # Merge and yield the function results, regardless of the termination status
+                # Include the ai_model_id so we can later add two streaming messages together
+                # Some settings may not have an ai_model_id, so we need to check for it
+                ai_model_id = self._get_ai_model_id(settings)
+                function_result_messages = merge_streaming_function_results(
+                    messages=chat_history.messages[-len(results) :],
+                    ai_model_id=ai_model_id,  # type: ignore
+                )
+                if self._yield_function_result_messages(function_result_messages):
+                    yield function_result_messages
```
Comment (eavanvalkenburg) [inline python/semantic_kernel/connectors/ai/chat_completion_client_base.py:318]: I like it, just wondering if this will break folks, if one has logic on what to do with FCC and TextContent and then suddenly you get this as well.. maybe we should add a flag on the class/call that is False by default and that decides whether to send or not.

File: python/semantic_kernel/contents/function_call_content.py [hunk 2]
```diff
@@ -149,7 +149,8 @@ def parse_arguments(self) -> Mapping[str, Any] | None:
         if isinstance(self.arguments, Mapping):
             return self.arguments
         try:
-            return json.loads(self.arguments)
+            sanitized_arguments = self.arguments.replace("'", '"')
+            return json.loads(sanitized_arguments)
```
Comment (eavanvalkenburg) [inline python/semantic_kernel/contents/function_call_content.py:153]:
```suggestion
            return json.loads(self.arguments.replace("'", '"'))
```

File: python/semantic_kernel/functions/kernel_function_from_prompt.py [hunk 3]
```diff
@@ -6,7 +6,7 @@
 from html import unescape
 from typing import TYPE_CHECKING, Any
 
-import yaml
+import yaml  # type: ignore
```
Comment (eavanvalkenburg) [inline python/semantic_kernel/functions/kernel_function_from_prompt.py:9]: why is this needed?

Comment (moonbox3) [reply python/semantic_kernel/functions/kernel_function_from_prompt.py:9]: Mypy throws an error

File: python/semantic_kernel/connectors/ai/chat_completion_client_base.py (hunk 1 above)
Comment (moonbox3) [reply python/semantic_kernel/connectors/ai/chat_completion_client_base.py:318]: I was thinking about if this is breaking, but the fact that it’s additive, had me thinking it would be okay. My thought was folks would be checking specific item types, so if wanting to handle this, they add a new case.

File: python/semantic_kernel/functions/kernel_function_from_prompt.py (hunk 3 above)
Comment (moonbox3) [reply python/semantic_kernel/functions/kernel_function_from_prompt.py:9]: image

**Comment by markwallace-microsoft** [comment]
<a href="https://github.com/microsoft/semantic-kernel/blob/0691fda7053ebd52de74a608f30074744e579bfb/README.md">Python Test Coverage</a><details><summary>Python Test Coverage Report • </summary><table><tr><th>File</th><th>Stmts</th><th>Miss</th><th>Cover</th><th>Missing</th></tr><tbody><tr><td colspan="5"><b>semantic_kernel</b></td></tr><tr><td>&nbsp; &nbsp;<a href="https://github.com/microsoft/semantic-kernel/blob/0691fda7053ebd52de74a608f30074744e579bfb/semantic_kernel/kernel.py">kernel.py</a></td><td>202</td><td>47</td><td>77%</td><td><a href="https://github.com/microsoft/semantic-kernel/blob/0691fda7053ebd52de74a608f30074744e579bfb/semantic_kernel/kernel.py#L143">143</a>, <a href="https://github.com/microsoft/semantic-kernel/blob/0691fda7053ebd52de74a608f30074744e579bfb/semantic_kernel/kernel.py#L154">154</a>, <a href="https://github.com/microsoft/semantic-kernel/blob/0691fda7053ebd52de74a608f30074744e579bfb/semantic_kernel/kernel.py#L158">158</a>, <a href="https://github.com/microsoft/semantic-kernel/blob/0691fda7053ebd52de74a608f30074744e579bfb/semantic_kernel/kernel.py#L308-L311">308&ndash;311</a>, <a href="https://github.com/microsoft/semantic-kernel/blob/0691fda7053ebd52de74a608f30074744e579bfb/semantic_kernel/kernel.py#L423">423</a>, <a href="https://github.com/microsoft/semantic-kernel/blob/0691fda7053ebd52de74a608f30074744e579bfb/semantic_kernel/kernel.py#L437-L480">437&ndash;480</a></td></tr><tr><td colspan="5"><b>semantic_kernel/connectors/ai</b></td></tr><tr><td>&nbsp; &nbsp;<a href="https://github.com/microsoft/semantic-kernel/blob/0691fda7053ebd52de74a608f30074744e579bfb/semantic_kernel/connectors/ai/chat_completion_client_base.py">chat_completion_client_base.py</a></td><td>127</td><td>2</td><td>98%</td><td><a href="https://github.com/microsoft/semantic-kernel/blob/0691fda7053ebd52de74a608f30074744e579bfb/semantic_kernel/connectors/ai/chat_completion_client_base.py#L401">401</a>, <a href="https://github.com/microsoft/semantic-kernel/blob/0691fda7053ebd52de74a608f30074744e579bfb/semantic_kernel/connectors/ai/chat_completion_client_base.py#L411">411</a></td></tr><tr><td colspan="5"><b>semantic_kernel/contents</b></td></tr><tr><td>&nbsp; &nbsp;<a href="https://github.com/microsoft/semantic-kernel/blob/0691fda7053ebd52de74a608f30074744e579bfb/semantic_kernel/contents/function_call_content.py">function_call_content.py</a></td><td>102</td><td>2</td><td>98%</td><td><a href="https://github.com/microsoft/semantic-kernel/blob/0691fda7053ebd52de74a608f30074744e579bfb/semantic_kernel/contents/function_call_content.py#L187">187</a>, <a href="https://github.com/microsoft/semantic-kernel/blob/0691fda7053ebd52de74a608f30074744e579bfb/semantic_kernel/contents/function_call_content.py#L215">215</a></td></tr><tr><td colspan="5"><b>semantic_kernel/functions</b></td></tr><tr><td>&nbsp; &nbsp;<a href="https://github.com/microsoft/semantic-kernel/blob/0691fda7053ebd52de74a608f30074744e579bfb/semantic_kernel/functions/kernel_function_from_prompt.py">kernel_function_from_prompt.py</a></td><td>154</td><td>7</td><td>95%</td><td><a href="https://github.com/microsoft/semantic-kernel/blob/0691fda7053ebd52de74a608f30074744e579bfb/semantic_kernel/functions/kernel_function_from_prompt.py#L167-L168">167&ndash;168</a>, <a href="https://github.com/microsoft/semantic-kernel/blob/0691fda7053ebd52de74a608f30074744e579bfb/semantic_kernel/functions/kernel_function_from_prompt.py#L182">182</a>, <a href="https://github.com/microsoft/semantic-kernel/blob/0691fda7053ebd52de74a608f30074744e579bfb/semantic_kernel/functions/kernel_function_from_prompt.py#L203">203</a>, <a href="https://github.com/microsoft/semantic-kernel/blob/0691fda7053ebd52de74a608f30074744e579bfb/semantic_kernel/functions/kernel_function_from_prompt.py#L221">221</a>, <a href="https://github.com/microsoft/semantic-kernel/blob/0691fda7053ebd52de74a608f30074744e579bfb/semantic_kernel/functions/kernel_function_from_prompt.py#L241">241</a>, <a href="https://github.com/microsoft/semantic-kernel/blob/0691fda7053ebd52de74a608f30074744e579bfb/semantic_kernel/functions/kernel_function_from_prompt.py#L324">324</a></td></tr><tr><td><b>TOTAL</b></td><td><b>16750</b></td><td><b>1849</b></td><td><b>89%</b></td><td>&nbsp;</td></tr></tbody></table></details>

## Python Unit Test Overview
| Tests | Skipped | Failures | Errors | Time |
| ----- | ------- | -------- | -------- | ------------------ |
| 2958 | 4 :zzz: | 0 :x: | 0 :fire: | 1m 24s :stopwatch: |

File: python/semantic_kernel/connectors/ai/chat_completion_client_base.py [hunk 4]
```diff
@@ -303,8 +306,18 @@ async def get_streaming_chat_message_contents(
                     ],
                 )
 
+                # Merge and yield the function results, regardless of the termination status
+                # Include the ai_model_id so we can later add two streaming messages together
+                # Some settings may not have an ai_model_id, so we need to check for it
+                ai_model_id = self._get_ai_model_id(settings)
```
Comment (TaoChenOSU) [inline python/semantic_kernel/connectors/ai/chat_completion_client_base.py:312]: The base class `AIServiceClientBase` also has a property `ai_model_id`, should we check there too?

File: python/semantic_kernel/connectors/ai/function_calling_utils.py [hunk 5]
```diff
@@ -95,3 +96,30 @@ def merge_function_results(
...
+        The combined streaming chat message content type.
+    """
+    items: list[Any] = []
+    for message in messages:
+        items.extend([item for item in message.items if isinstance(item, FunctionResultContent)])
+
+    # If we want to be able to support adding the streaming message chunks together, then the author role needs to be
+    # `Assistant```, as the `Tool` role will cause the add method to break.
+    return [
+        StreamingChatMessageContent(role=AuthorRole.ASSISTANT, items=items, choice_index=0, ai_model_id=ai_model_id)
```
Comment (TaoChenOSU) [inline python/semantic_kernel/connectors/ai/function_calling_utils.py:124]: This is going to cause issue in some connectors if the function result is not with a role that they expect. For example, if someone uses a connector at the beginning then later switch to another to continue the same chat history.

Why couldn't streaming messages with the tool role be added together?

File: python/semantic_kernel/functions/kernel_function_from_prompt.py (hunk 3 above)
Comment (TaoChenOSU) [reply python/semantic_kernel/functions/kernel_function_from_prompt.py:9]: Seems related: https://github.com/python/mypy/issues/10632

File: python/semantic_kernel/kernel.py [hunk 6]
```diff
@@ -398,7 +399,12 @@ async def invoke_function_call(
         frc = FunctionResultContent.from_function_call_content_and_result(
             function_call_content=function_call, result=invocation_context.function_result
         )
-        chat_history.add_message(message=frc.to_chat_message_content())
+
+        is_streaming = any(isinstance(message, StreamingChatMessageContent) for message in chat_history.messages)
```
Comment (TaoChenOSU) [inline python/semantic_kernel/kernel.py:403]: There are two more places in this method where you'd need to check if the chat is carried out in a streaming manner.

File: python/tests/unit/connectors/ai/azure_ai_inference/services/test_azure_ai_inference_chat_completion.py [hunk 7]
```diff
@@ -507,20 +509,31 @@ async def test_azure_ai_inference_streaming_chat_completion_with_function_choice
...
-        assert messages[0].finish_reason == FinishReason.TOOL_CALLS
+        all_messages.extend(messages)
+
+    # Assert the number of total messages
+    assert len(all_messages) == 2, f"Expected 2 messages, got {len(all_messages)}"
+
+    # Validate the first message
+    assert all_messages[0].role == "assistant", f"Unexpected role for first message: {all_messages[0].role}"
+    assert all_messages[0].content == "", f"Unexpected content for first message: {all_messages[0].content}"
+    assert all_messages[0].finish_reason == FinishReason.TOOL_CALLS, (
```
Comment (TaoChenOSU) [inline python/tests/unit/connectors/ai/azure_ai_inference/services/test_azure_ai_inference_chat_completion.py:529]: Why is the finish_reason of the first one `TOOL_CALLS`? The mock has the second message with the `TOOL_CALLS` as the finish reason.

File: python/tests/unit/connectors/ai/google/google_ai/services/test_google_ai_chat_completion.py [hunk 8]
```diff
@@ -275,20 +277,29 @@ async def test_google_ai_streaming_chat_completion_with_function_choice_behavior
...
         kernel=kernel,
     ):
-        assert len(messages) == 1
-        assert messages[0].role == "assistant"
-        assert messages[0].content == ""
-        # Google doesn't return STOP as the finish reason for tool calls
-        assert messages[0].finish_reason == FinishReason.STOP
+        all_messages.extend(messages)
+
+    assert len(all_messages) == 2, f"Expected 2 messages, got {len(all_messages)}"
```
Comment (TaoChenOSU) [inline python/tests/unit/connectors/ai/google/google_ai/services/test_google_ai_chat_completion.py:290]: Why will there be 2 messages? The mock only has one.

File: python/tests/unit/connectors/ai/google/vertex_ai/services/test_vertex_ai_chat_completion.py [hunk 9]
```diff
@@ -275,20 +277,29 @@ async def test_vertex_ai_streaming_chat_completion_with_function_choice_behavior
...
         kernel=kernel,
     ):
-        assert len(messages) == 1
-        assert messages[0].role == "assistant"
-        assert messages[0].content == ""
-        # Google doesn't return STOP as the finish reason for tool calls
-        assert messages[0].finish_reason == FinishReason.STOP
+        all_messages.extend(messages)
+
+    assert len(all_messages) == 2, f"Expected 2 messages, got {len(all_messages)}"
```
Comment (TaoChenOSU) [inline python/tests/unit/connectors/ai/google/vertex_ai/services/test_vertex_ai_chat_completion.py:290]: Same question as above.

File: python/semantic_kernel/connectors/ai/chat_completion_client_base.py (hunk 1 above)
Comment (TaoChenOSU) [reply python/semantic_kernel/connectors/ai/chat_completion_client_base.py:318]: This is probably breaking but the impact is small and it's something we have to fix.

File: python/semantic_kernel/connectors/ai/function_calling_utils.py (hunk 5 above)
Comment (moonbox3) [reply python/semantic_kernel/connectors/ai/function_calling_utils.py:124]: It seems there was a conscious decision to not allow to streaming chat message content chunks to be added together if their roles are different: https://github.com/microsoft/semantic-kernel/blob/62a50f32cf140b24876517219726a28465ef640e/python/semantic_kernel/contents/streaming_chat_message_content.py#L171

File: python/semantic_kernel/kernel.py (hunk 6 above)
Comment (moonbox3) [reply python/semantic_kernel/kernel.py:403]: I don’t follow?

File: python/tests/unit/connectors/ai/azure_ai_inference/services/test_azure_ai_inference_chat_completion.py (hunk 7 above)
Comment (moonbox3) [reply python/tests/unit/connectors/ai/azure_ai_inference/services/test_azure_ai_inference_chat_completion.py:529]: The first finish reason is related to the FunctionCallContent.

File: python/tests/unit/connectors/ai/google/google_ai/services/test_google_ai_chat_completion.py (hunk 8 above)
Comment (moonbox3) [reply python/tests/unit/connectors/ai/google/google_ai/services/test_google_ai_chat_completion.py:290]: This test was broken before. During kernel function invocation it errored out while trying to create the json args because they contained single quotes and not double. The error was never caught or handled because it’s a message added to the chat history. 

This test has two messages now: yielding a function call content message and now with the addition in the PR, a function result content message.

File: python/tests/unit/connectors/ai/google/vertex_ai/services/test_vertex_ai_chat_completion.py (hunk 9 above)
Comment (moonbox3) [reply python/tests/unit/connectors/ai/google/vertex_ai/services/test_vertex_ai_chat_completion.py:290]: See above.

File: python/semantic_kernel/connectors/ai/chat_completion_client_base.py (hunk 4 above)
Comment (moonbox3) [reply python/semantic_kernel/connectors/ai/chat_completion_client_base.py:312]: Definitely, thanks for the callout. Updated to fallback to the `self.ai_model_id` if `settings.ai_model_id` doesn't exist (or is blank). (Side note: it's interesting that our `ai_model_id` in the prompt execution settings allows or "" strings.)

File: python/semantic_kernel/kernel.py (hunk 6 above)
Comment (moonbox3) [reply python/semantic_kernel/kernel.py:403]: Are you referring to these: `message=frc.to_chat_message_content()`? I don't think it matters that those "error" messages are converted to the Streaming CMC. The fact that we have a Streaming CMC as a response from the model, should indicate enough that we are streaming -- I was doing this check to avoid having to introduce a new `is_streaming` kwarg to the invoke_function_call method.

Comment (moonbox3) [reply python/semantic_kernel/kernel.py:403]: Now that I think about it, too: if we converted those intermittent "error" messages to Streaming CMC, then if someone does combine the streaming chat message chunks together, they would see those "hey model, please reformat the json and try calling again" messages...

File: python/tests/unit/connectors/ai/azure_ai_inference/services/test_azure_ai_inference_chat_completion.py (hunk 7 above)
Comment (TaoChenOSU) [reply python/tests/unit/connectors/ai/azure_ai_inference/services/test_azure_ai_inference_chat_completion.py:529]: Oh, I see it has an empty list of choices thus the first message was skipped.

File: python/semantic_kernel/connectors/ai/chat_completion_client_base.py (hunk 4 above)
Comment (TaoChenOSU) [reply python/semantic_kernel/connectors/ai/chat_completion_client_base.py:312]: I don't see the update. Am I missing something?

Comment (moonbox3) [reply python/semantic_kernel/connectors/ai/chat_completion_client_base.py:312]: I haven't pushed it yet. Will push now.

File: python/semantic_kernel/kernel.py (hunk 6 above)
Comment (TaoChenOSU) [reply python/semantic_kernel/kernel.py:403]: Another question is: when do we need to concatenate the tool result messages?

File: python/semantic_kernel/connectors/ai/function_calling_utils.py (hunk 5 above)
Comment (TaoChenOSU) [reply python/semantic_kernel/connectors/ai/function_calling_utils.py:124]: That's ok, no? These are tool messages; we don't want them to be concatenated with the assistant message.

File: python/samples/concepts/auto_function_calling/chat_gpt_api_function_calling.py [hunk 10]
```diff
@@ -130,13 +131,15 @@ async def handle_streaming(
...
-        if not execution_settings.function_choice_behavior.auto_invoke_kernel_functions and isinstance(
-            message[0], StreamingChatMessageContent
+        if (
+            not execution_settings.function_choice_behavior.auto_invoke_kernel_functions
+            and isinstance(message[0], StreamingChatMessageContent)
+            and message[0].role == AuthorRole.ASSISTANT
         ):
             streamed_chunks.append(message[0])
-        else:
+        elif isinstance(message[0], StreamingChatMessageContent) and message[0].role == AuthorRole.ASSISTANT:
```
Comment (TaoChenOSU) [inline python/samples/concepts/auto_function_calling/chat_gpt_api_function_calling.py:142]: Question: why are these conditions necessary?

File: python/semantic_kernel/connectors/ai/function_calling_utils.py [hunk 11]
```diff
@@ -120,6 +120,4 @@ def merge_streaming_function_results(
 
     # If we want to be able to support adding the streaming message chunks together, then the author role needs to be
     # `Assistant```, as the `Tool` role will cause the add method to break.
```
Comment (TaoChenOSU) [inline python/semantic_kernel/connectors/ai/function_calling_utils.py:122]: The comment is not updated.

File: python/semantic_kernel/kernel.py (hunk 6 above)
Comment (moonbox3) [reply python/semantic_kernel/kernel.py:403]: We have examples of concatenating other messages, so this would be an example on how to do that.

File: python/semantic_kernel/connectors/ai/function_calling_utils.py (hunk 5 above)
Comment (moonbox3) [reply python/semantic_kernel/connectors/ai/function_calling_utils.py]: Synced offline - will keep the Streaming CMC for FRC as a TOOL role.

File: python/semantic_kernel/connectors/ai/function_calling_utils.py (hunk 11 above)
Comment (moonbox3) [reply python/semantic_kernel/connectors/ai/function_calling_utils.py:122]: Removed comment as it no longer applies.

File: python/samples/concepts/auto_function_calling/chat_gpt_api_function_calling.py (hunk 10 above)
Comment (moonbox3) [reply python/samples/concepts/auto_function_calling/chat_gpt_api_function_calling.py:142]: As we've been talking about -- if we simply leave the code as is, it will break. That's because all streaming CMC results are added to one list. We only want to add the Assistant message roles to this list, so that we can later form one CMC.

**Review by TaoChenOSU — APPROVED** [review]
